src/dsp_tools/utils/xmlupload/list_client.py

```diff
@@ -0,0 +1,120 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Any, Iterable, Protocol
+from urllib.parse import quote_plus
+
+import requests
+
+from dsp_tools.models.exceptions import BaseError
+from dsp_tools.utils.create_logger import get_logger
+
+# pylint: disable=too-few-public-methods
+
+logger = get_logger(__name__)
+
+
+@dataclass(frozen=True)
+class ListNode:
+    """Information about a list node."""
+
+    node_iri: str
+    node_name: str
+
+
+@dataclass(frozen=True)
+class List:
+    """Information about a list."""
+
+    root_iri: str
+    list_name: str
+    children: list[ListNode]
+
+
+@dataclass(frozen=True)
+class ProjectLists:
+    """Information about the lists of a project."""
+
+    lists: list[List]
+
+
+class ListClient(Protocol):
+    """Interface (protocol) for list-related requests to the DSP-API."""
+
+    def get_list_node_id_to_iri_lookup(self) -> dict[str, str]:
+        """Get a lookup of list node names to IRIs."""
+
+
+@dataclass
+class ListClientLive:
+    """Client handling list-related requests to the DSP-API."""
+
+    server: str
+    project_iri: str
+    list_info: ProjectLists | None = field(init=False, default=None)
+
+    def get_list_node_id_to_iri_lookup(self) -> dict[str, str]:
+        """
+        Get a mapping of list node IDs to their respective IRIs.
+        A list node ID is structured as follows:
+        <list name>:<node name> where the list name is the node name of the root node.
+
+        Returns:
+            The mapping of list node IDs to IRIs.
+        """
+        if not self.list_info:
+            self.list_info = _get_list_info_from_server(self.server, self.project_iri)
+        return dict(_get_node_tuples(self.list_info.lists))
+
+
+def _get_node_tuples(lists: list[List]) -> Iterable[tuple[str, str]]:
+    for l in lists:
+        list_name = l.list_name
+        for node in l.children:
+            node_name = node.node_name
+            node_id = f"{list_name}:{node_name}"
+            yield node_id, node.node_iri
+
+
+def _get_list_info_from_server(server: str, project_iri: str) -> ProjectLists:
+    logger.info(f"Retrieving lists of project {project_iri}")
+    list_iris = _get_list_iris_from_server(server, project_iri)
+    lists = [_get_list_from_server(server, list_iri) for list_iri in list_iris]
+    return ProjectLists(lists)
+
+
+def _get_list_iris_from_server(server: str, project_iri: str) -> list[str]:
+    iri = quote_plus(project_iri)
+    res = requests.get(f"{server}/admin/lists?projectIri={iri}", timeout=5)
```

Comment:
I'm a bit torn about adding two layers of indirection to each request - simply because it makes bug hunting harder. But I see the benefits and will implement it like this. In the long term we may even be able to get rid of the connection altogether? But let's discuss that some other time.

I assume the same goes for the ProjectClient too?

Reply:
Yes, I would generally always use try_network_action + connection class.